### PR TITLE
63 Countdown

### LINKED ISSUE #63: Countdown
Countdown for when the game starts (or resumes)

### PR DESCRIPTION
**Pull request by thervh70**
Will close #63 

This is fully implemented and the E2E test is adapted to the new
countdown.

< feature

### CODE REVIEW

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java [hunk 1]
```diff
@@ -214,5 +226,59 @@ private static PhongMaterial chooseMaterial() {
...
+		countdown.setLayoutY(ShaftEscape.HEIGHT / 2 - offsetY);
+
+		countdownAnimation(index);
+	}
+
+	/**
+	 * Render the animations for the countdown method.
+	 * @param index The index to be rendered.
+	 */
+	private void countdownAnimation(final Integer index) {
```
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: *Auto-generated comment*

    Method length is 31 lines (max allowed is 30).

File: src/test/java/nl/tudelft/ti2206/group9/gui/EndToEndTest.java [hunk 2]
```diff
@@ -78,8 +82,10 @@ public void test() throws IOException {
 
 		mainMenu(0);				// Click start
 		playerDies();				// Player dies
+		sleep(COUNTDOWN);
 		deathPopup(0);				// Click "Try Again"
```
Comment (mpsijm) [inline src/test/java/nl/tudelft/ti2206/group9/gui/EndToEndTest.java]: In my opinion, the `sleep(COUNTDOWN)` should be right after `mainMenu(0)`. In your code, the player dies before the countdown has ended.

<logic,review

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java [hunk 3]
```diff
@@ -30,6 +39,8 @@
 	/** Width of the box in-game where the score is displayed. */
 	private static final int SCORE_BOX_WIDTH = 140;
 
+	private Label countdown = new Label();
```
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: *Auto-generated comment*

    Missing a Javadoc comment.

**Comment by mpsijm** [comment]
Please, run PMD and Checkstyle again on your project. There are quite a lot of PMD and Checkstyle warnings left accordingin to my IDE.
Also, I have found a small glitch: pressing Resume in the pause popup will still render one frame before starting the countdown. Maybe it renders the last frame that should have been rendered when the pause popup emerged, because the InternalTicker hasn't been started yet during/before the countdown.

<bug,layout,documentation

**Comment by thervh70** [comment]
@mpsijm According to my IDE my branch is "clean".
If not, in that case i need help with setting up my IDE.

< review, configuration

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java [hunk 4]
```diff
@@ -214,5 +226,60 @@ private static PhongMaterial chooseMaterial() {
...
+		countdown.setLayoutY(ShaftEscape.HEIGHT / 2 - offsetY);
+
+		countdownAnimation(index);
+	}
+
+	/**
+	 * Render the animations for the countdown method.
+	 * @param index The index to be rendered.
+	 */
+	private void countdownAnimation(
```
Comment (thervh70) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: *Auto-generated comment*

    Method length is 31 lines (max allowed is 30).

**Comment by mpsijm** [comment]
I can help you if you need help setting up your IDE, however I have no experience with IntelliJ. @MathiasMeuleman or @MKerkhof should be able to give you all the button clicks needed to set everything up :)

<review,configuration

**Comment by MKerkhof** [comment]
Using our own PMD and checkstyle rules, I still get PMD violations. See the image for a more precise overview of the violations and there location. 

capture

As for checkstyle, 0 violations :)

<review

**Comment by thervh70** [comment]
Please review, i have merged the most recent master into my branch.

**Comment by MathiasMeuleman** [comment]
I fixed your PMD, now it should be mergable :)

<review, organisation

**Comment by MKerkhof** [comment]
Looks good, nice job on resolving all warnings. Functionality is good too. Overall, looks ready to merge! The next person that reviews it can also merge it in my opninion!

<review

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java [hunk 5]
```diff
@@ -214,5 +228,59 @@ private static PhongMaterial chooseMaterial() {
...
+				Duration.millis(duration), countdownLabel);
+		ft.setAutoReverse(true);
+		ft.setCycleCount(2);
+		ft.setToValue(1);
+		ft.setFromValue(0);
+		ft.play();
+
+		st.setOnFinished(new EventHandler<ActionEvent>() {
+			public void handle(final ActionEvent event) {
+				if (index > INT_ONE) {
```
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: Does PMD really warn about 1 being a magic number?

<review,configuration

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java [hunk 6]
```diff
@@ -214,5 +228,59 @@ private static PhongMaterial chooseMaterial() {
 			return Style.MOSS;
 		}
 	}
+
+	/**
+	 * Render the countdownLabel method.
+	 * @param index The index to be rendered.
+	 */
+	public void countdown(final Integer index) {
```
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: Why did you use `Integer` here, instead of `int`?

<review,naming

**Comment by mpsijm** [comment]
The feature looks very good, and it works perfectly as expected. I do have some questions about small things in the code, see the line comments.

<review,feature

**Comment by mdingjan** [comment]
A nice addition to the current game :) I reviewed your changes, these are fine.
Please answer the questions above. When you've done that, @mpsijm can merge this branch if he is satisfied with your answer. ;)

<review

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java (hunk 5 above)
Comment (MathiasMeuleman) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: Yes he does and I don't like him for it! I believe it is only in conditional statements, but still...

<review, configuration

Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: I think I get what you mean. I hope that you're able to refactor this to something without the `INT_ONE` constant. Maybe `if (index - 1 > 0)`?

<review, logic, configuration

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java (hunk 6 above)
Comment (thervh70) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: To be able to use index.toString() in line 241.

< naming

Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: Then I would propose using `Integer.toString(int)` instead.

<review, naming

File: src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java (hunk 5 above)
Comment (mpsijm) [inline src/main/java/nl/tudelft/ti2206/group9/gui/ExternalTicker.java]: Other suggestion:

    int newIndex = index - 1;
    if (newIndex > 0) {
        countdown(newIndex);
    } else {
        ...
    }

<review, logic

**Comment by mpsijm** [comment]
Thanks for processing my comments :) Ready for take-off! I mean, merge! 

<review

**Pull request merged**